org/mozilla/jss/ssl/javax/JSSEngineReferenceImpl.java

```diff
@@ -0,0 +1,1170 @@
+package org.mozilla.jss.ssl.javax;
+
+import java.lang.*;
+import java.util.*;
+
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.nio.channels.WritableByteChannel;
+import java.nio.channels.Channels;
+
+import java.nio.ByteBuffer;
+
+import javax.net.ssl.*;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.mozilla.jss.nss.*;
+import org.mozilla.jss.pkcs11.*;
+import org.mozilla.jss.provider.javax.crypto.*;
+import org.mozilla.jss.ssl.*;
+
+import org.mozilla.jss.crypto.Policy;
+import org.mozilla.jss.crypto.PrivateKey;
+import org.mozilla.jss.crypto.X509Certificate;
+
+/**
+ * The reference JSSEngine implementation.
+ *
+ * This JSSEngine implementation is a reference for future JSSEngine
+ * implementations, providing a pure-Java overview of what should happen at
+ * each step of the init, wrap, and unwrap calls.
+ *
+ * This implementation allows for extended debug logging, but also debug
+ * packet logging. The latter writes out packets sent via wrap(...) and
+ * received from unwrap(...) to a port on localhost. This allows one to easily
+ * attach Wireshark or tcpdump and inspect the TLS packets, even if errors
+ * occur during the test suite (where packets aren't sent over the wire by
+ * default). This maintains the client/server relationship, and are logged
+ * as being from the appropriate side of the TLS connection.
+ */
+public class JSSEngineReferenceImpl extends JSSEngine {
+    private String peer_info;
+
+    private boolean closed_fd;
+    private BufferProxy read_buf;
+    private BufferProxy write_buf;
+    private SSLFDProxy ssl_fd;
+
+    private int unknown_state_count;
+    private boolean step_handshake;
+
+    private SSLException ssl_exception;
+    private boolean seen_exception;
+
+    private int debug_port;
+    private ServerSocket ss_socket;
+    private Socket s_socket;
+    private Socket c_socket;
+    private OutputStream s_ostream;
+    private OutputStream c_ostream;
+
+    private String name;
+    private String prefix = "";
+
+    public JSSEngineReferenceImpl() {
+        super();
+
+        peer_info = "";
+
+        debug("JSSEngine: constructor()");
+    }
+
+    public JSSEngineReferenceImpl(String peerHost, int peerPort) {
+        super(peerHost, peerPort);
+
+        peer_info = peerHost + ":" + peerPort;
+
+        debug("JSSEngine: constructor(" + peerHost + ", " + peerPort + ")");
```

Comment:
Eh, I'm not sure it matters much. One of the configuration options for the logger lets you see the actual class name that reported the exception too. I think it is disabled by default.

Maybe @edewata has a preference?

Reply:
It's really up to you. For troubleshooting purposes it would be better to match the log message with the class name (the logger can also insert the class name automatically depending on the logging config), but if the class is never meant to be exposed or there can be only one implementation at runtime that's probably ok.